log4j-layout-template-json/src/test/java/org/apache/logging/log4j/layout/template/json/GcpLayoutTest.java

```diff
@@ -0,0 +1,163 @@
+import org.apache.logging.log4j.util.SortedArrayStringMap;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.time.Instant;
+import java.util.*;
+
+import static org.apache.logging.log4j.layout.template.json.TestHelpers.serializeUsingLayout;
+
+class GcpLayoutTest {
+
+    private static final Configuration CONFIGURATION = new DefaultConfiguration();
+
+    private static final Charset CHARSET = StandardCharsets.UTF_8;
+
+    private static final String SERVICE_NAME = "test";
+
+    private static final JsonTemplateLayout JSON_TEMPLATE_LAYOUT = JsonTemplateLayout
+            .newBuilder()
+            .setConfiguration(CONFIGURATION)
+            .setCharset(CHARSET)
+            .setEventTemplateUri("classpath:GcpLayout.json")
+            .build();
+
+    @Test
+    void test_lite_log_events() {
+        final List<LogEvent> logEvents = LogEventFixture.createLiteLogEvents(1_000);
+        test(logEvents);
+    }
+
+    @Test
+    void test_full_log_events() {
+        final List<LogEvent> logEvents = LogEventFixture.createFullLogEvents(1_000);
+        test(logEvents);
+    }
+
+    private static void test(final Collection<LogEvent> logEvents) {
+        for (final LogEvent logEvent : logEvents) {
+            test(logEvent);
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    private static void test(final LogEvent logEvent) {
+        final Map<String, Object> jsonTemplateLayoutMap = renderUsingJsonTemplateLayout(logEvent);
+        Assertions.assertThat(Instant.parse(jsonTemplateLayoutMap.get("timestamp").toString()).toEpochMilli())
+          .isEqualTo(logEvent.getTimeMillis());
+        Assertions.assertThat(jsonTemplateLayoutMap.get("severity")).isEqualTo(logEvent.getLevel().toString());
+        Assertions.assertThat(jsonTemplateLayoutMap.get("_logger")).isEqualTo(logEvent.getLoggerName());
+        Assertions.assertThat(jsonTemplateLayoutMap.get("_thread").toString())
+          .isEqualTo(logEvent.getThreadName());
+
+        if (logEvent.getContextData().size() == 0) {
+            Assertions.assertThat(jsonTemplateLayoutMap.get("logging.googleapis.com/labels"))
+              .isNull();
+        } else {
+            Assertions.assertThat(logEvent.getContextData().toMap().keySet())
+              .isEqualTo(((Map)jsonTemplateLayoutMap.get("logging.googleapis.com/labels")).keySet());
+            logEvent.getContextData().toMap().keySet().forEach(k -> {
+                Assertions.assertThat((Object) logEvent.getContextData().getValue(k))
+                  .isEqualTo(((Map)jsonTemplateLayoutMap.get("logging.googleapis.com/labels")).get(k));
+            });
+        }
+
+        if (logEvent.getSource() == null) {
+            // not quite sure why the output has "?." in sourceLocation.function
```

Comment:
This feels like a bug (feature?) in `PatternLayout`. I am checking this out...

Reply:
What is going on is this: `ClassNamePatternConverter` emits `ClassNamePatternConverter.NA` (which is `?`) due to the missing source location, next `PatternLayout` emits the `.` literal, finally `MethodLocationPatternConverter` emits nothing. I guess we need to implement `ConcatenationResolver` addressed in [LOG4J2-3120](https://issues.apache.org/jira/browse/LOG4J2-3120) first. I will also check `PatternLayout` documentation if we can replace `?.` outputs with `null`.

Reply:
`%replace{%C.%M}{^\?\.$}{}` does the trick, sort of, for now.